api/src/wfl/sink.clj

```diff
@@ -317,27 +325,63 @@
+   (letfn [(literal? [x] (str/starts-with? x "$"))
+           (datatype [k] (let [columns (->> schema
+                                            :tables
+                                            (filter #(= table (:name %)))
+                                            first
+                                            :columns)]
+                           (-> (filter #(= k (:name %)) columns)
+                               first
+                               :datatype)))
+           (fileref? [k] (= (datatype (name k)) "fileref"))
+           (boolean? [k] (= (datatype (name k)) "boolean"))
+           (get-target [url]
+             (let [[bucket obj] (storage/parse-gs-url url)]
+               (if (= target-bucket "")
+                 (storage/gs-url bucket (str/join "/" [workflow-id (util/basename obj)]))
+                 (str target-bucket workflow-id "/" (util/basename obj)))))
+           (go! [k v]
+             (cond (fileref? k) (let [val (values (keyword v))]
+                                  [k {:description (util/basename val)
+                                      :mimeType   (mime-type/ext-mime-type val)
+                                      :sourcePath val
+                                      :targetPath (get-target val)}])
+                   (literal? v) [k (subs v 1 (count v))]
+                   (boolean? v) [k (values (keyword v))]
+                   (string?  v) [k (values (keyword v))]
+                   (map?     v) [k (rename-gather-bulk workflow-id dataset table values v)]
+                   (coll?    v) [k (keep (go! k v))]
+                   :else        (throw (ex-info "Unknown operation"
+                                                {:operation v}))))]
+     (into {} (for [[k v] mapping] (go! k v))))))
+
 (defn ^:private to-dataset-row
-  "Use `fromOutputs` to coerce the `workflow` outputs into a row in the dataset
-   where `fromOutputs` describes a mapping from workflow outputs to columns in
-   the dataset table."
-  [fromOutputs {:keys [outputs] :as workflow}]
+  "Use `fromOutputs` and the schema provided in the `table` within the `dataset` to coerce the `workflow` outputs into a row in the dataset table. The dataset table schema describes column types while `fromOutputs` provides a mapping between workflow outputs and the table."
+  [dataset table fromOutputs {:keys [outputs uuid] :as workflow}]
   (when-not (map? fromOutputs)
     (throw (IllegalStateException. "fromOutputs is malformed")))
   (try
-    (rename-gather outputs fromOutputs)
+    (rename-gather-bulk uuid dataset table outputs fromOutputs)
     (catch Exception cause
       (throw (ex-info "Failed to coerce workflow outputs to dataset columns"
-                      {:fromOutputs fromOutputs :workflow workflow}
+                      {:outputs outputs :table table :fromOutputs fromOutputs :workflow workflow :dataset dataset}
                       cause)))))
```

Comment:
Perhaps just dump the sink itself here?
What are we leaving out of the `ex-data`?
Is that intentional?